// mjml-core/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerElement = exports.mjml2html = exports.version = exports.documentParser = exports.MJMLHeadElements = exports.registerMJHeadElement = exports.helpers = exports.elements = exports.registerMJElement = exports.MJMLRenderer = exports.MJMLElement = undefined;

var _warning = require('warning');

var _warning2 = _interopRequireDefault(_warning);

var _MJMLRenderer = require('./MJMLRenderer');

var _MJMLRenderer2 = _interopRequireDefault(_MJMLRenderer);

var _MJMLElementsCollection = require('./MJMLElementsCollection');

var _MJMLElementsCollection2 = _interopRequireDefault(_MJMLElementsCollection);

var _MJMLHead = require('./MJMLHead');

var _MJMLHead2 = _interopRequireDefault(_MJMLHead);

var _helpers = require('./helpers');

var helpers = _interopRequireWildcard(_helpers);

var _MJMLElement2 = require('./decorators/MJMLElement');

var _MJMLElement3 = _interopRequireDefault(_MJMLElement2);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.MJMLElement = _MJMLElement3.default;
exports.MJMLRenderer = _MJMLRenderer2.default;
exports.registerMJElement = _MJMLElementsCollection.registerMJElement;
exports.elements = _MJMLElementsCollection2.default;
exports.helpers = helpers;
exports.registerMJHeadElement = _MJMLHead.registerMJHeadElement;
exports.MJMLHeadElements = _MJMLHead2.default;
var documentParser = exports.documentParser = function documentParser(content) {
  var documentParser = require('./parsers/document').default;

  return documentParser(content);
};
var version = exports.version = function version() {
  return '2.3.3';
};
var mjml2html = exports.mjml2html = function mjml2html(mjml) {
  var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];
  return new _MJMLRenderer2.default(mjml, options).render();
};
var registerElement = exports.registerElement = function registerElement(Component) {
  (0, _warning2.default)(false, 'Please now use registerMJElement, registerElement is deprecated will no longer be supported soon');

  return (0, _MJMLElementsCollection.registerMJElement)(Component);
};
